### PR TITLE
Optimization for llm/gpt-3

### PR DESCRIPTION
**Pull request by DrownFish19**
### PR types

Function optimization

### PR changes

APIs and Docs
### Description

Optimization for llm/gpt-3
1. update `README.md` file.
2. replace `self.` parameters with `config.` parameters in `modeling.py`.
3. add `_init_weights` for `GPTPretrainedModel`.
4. modify `output_attentions` (`need_weights`) to control attention weights output.

### CODE REVIEW

File: llm/gpt-3/modeling.py [hunk 1]
```diff
@@ -98,45 +98,38 @@ class MultiHeadAttention(nn.Layer):
...
-        recompute_granularity="full",
-        do_recompute=True,
-        use_flash_attn=False,
+        config,
+        kdim=None, #
+        vdim=None, #
+        need_weights=False, # 
+        weight_attr=None, #
+        bias_attr=None, #
+        do_recompute=False,
```
Comment (ZHUI) [inline llm/gpt-3/modeling.py]: 看一下这些 参数的使用，应该都可以删除

```python
        kdim=None, #
        vdim=None, #
        need_weights=False, # 
        weight_attr=None, #
        bias_attr=None, #
        do_recompute=False,
```

File: llm/gpt-3/modeling.py [hunk 2]
```diff
@@ -98,45 +98,38 @@ class MultiHeadAttention(nn.Layer):
...
-        self.kdim = kdim if kdim is not None else embed_dim
-        self.vdim = vdim if vdim is not None else embed_dim
-        self.num_heads = num_heads
-        self.dropout = dropout
+        
+        self.config = config
+        embed_dim = config.hidden_size
+        self.embed_dim = config.hidden_size
+        self.kdim = kdim if kdim is not None else config.hidden_size
+        self.vdim = vdim if vdim is not None else config.hidden_size
```
Comment (ZHUI) [inline llm/gpt-3/modeling.py]: 这个 kdim vdim 应该不单独传入，直接 用 config.hidden_size

File: llm/gpt-3/modeling.py (hunk 1 above)
Comment (DrownFish19) [reply llm/gpt-3/modeling.py]: 删除了kdim, vdim, need_weights和bias_attr, 保留了weight_attr和do_recompute作为TransformerDecoderLayer参数接口

File: llm/gpt-3/modeling.py (hunk 2 above)
Comment (DrownFish19) [reply llm/gpt-3/modeling.py]: 这两处都直接使用hidden_size了，删除了kdim和vdim

File: llm/gpt-3/modeling.py (hunk 1 above)
Comment (ZHUI) [reply llm/gpt-3/modeling.py]: https://github.com/PaddlePaddle/PaddleNLP/pull/6403/files#diff-d7382bad9889b21de808d9a410c4eb01aa833043bdb611bc2e40d3ded1878846

File: llm/gpt-3/modeling.py [hunk 3]
```diff
@@ -98,45 +98,35 @@ class MultiHeadAttention(nn.Layer):
...
+        self.head_dim = embed_dim // self.num_heads
+        assert self.head_dim * self.num_heads == self.embed_dim, "embed_dim must be divisible by num_heads"
 
-        assert self.num_heads % num_partitions == 0
-        self.num_heads = self.num_heads // num_partitions
+        assert self.num_heads % config.tensor_parallel_degree == 0
+        self.num_heads = self.num_heads // config.tensor_parallel_degree
 
-        if num_partitions > 1:
+        if config.tensor_parallel_degree > 1:
```
Comment (ZHUI) [inline llm/gpt-3/modeling.py:113]: assert self.num_heads % config.tensor_parallel_degree == 0
        self.num_heads = self.num_heads // config.tensor_parallel_degree

File: llm/gpt-3/modeling.py [hunk 4]
```diff
@@ -598,31 +577,45 @@ def __init__(
...
 
+        self.apply(self._init_weights)
+
+    def _init_weights(self, layer):
+        """Initialization Weights"""
+
+        if isinstance(layer, (nn.Linear, 
+                              nn.Embedding,
+                              fleet.meta_parallel.VocabParallelEmbedding)):
+            # In the dygraph mode, use the `set_value` to reset the parameter directly,
```
Comment (ZHUI) [inline llm/gpt-3/modeling.py]: 此处不全，参考llama

File: llm/gpt-3/modeling.py [hunk 5]
```diff
@@ -682,6 +669,17 @@ def get_tensor_parallel_split_mappings(num_layers):
                 "layers.0.linear2.weight": partial(fn, is_column=False),
             }
 
+            if config.fuse_attention_qkv:
+                base_actions["layers.0.self_attn.qkv_proj.weight"] = partial(fn, is_column=True)
```
Comment (ZHUI) [inline llm/gpt-3/modeling.py:599]: 单卡。tp=2前向精度。

Comment (ZHUI) [reply llm/gpt-3/modeling.py:599]: https://github.com/PaddlePaddle/PaddleNLP/blob/3b8569206d23a8bbf615ae3d2a3862c4bb12a7b7/llm/glm/tests/test_glm_mp.py#L111C1-L113C35

File: llm/gpt-3/modeling.py [hunk 6]
```diff
@@ -98,45 +98,35 @@ class MultiHeadAttention(nn.Layer):
...
+        self.fuse_attention_qkv = config.fuse_attention_qkv
+        self.scale_qk_coeff = getattr(config,"scale_qk_coeff", 1.0)
+        self.use_recompute = config.use_recompute
+        self.recompute_granularity = getattr(config,"recompute_granularity", "full")
         self.do_recompute = do_recompute
-        self.use_flash_attn = use_flash_attn if flash_attention else None
-
-        self.head_dim = embed_dim // num_heads
-        assert self.head_dim * num_heads == self.embed_dim, "embed_dim must be divisible by num_heads"
+        self.use_flash_attn = config.use_flash_attn if flash_attention else None
```
Comment (ZHUI) [inline llm/gpt-3/modeling.py]: use_flash_attention

File: llm/gpt-3/README.md [hunk 7]
```diff
@@ -28,15 +30,20 @@ mv gpt_en_dataset_300m_ids.npy ./data
 mv gpt_en_dataset_300m_idx.npz ./data
 ```
 
-使用下面脚本,即可在llama-7b的基础上,继续训练.
+注意：
+1. 需要paddle develop版本训练，需要安装`pip install tool_helpers visualdl==2.5.3`等相关缺失whl包
+2. `use_flash_attn` 需要在A100机器开启，否则loss可能不正常（很快变成0.00x,非常小不正常）。建议使用cuda11.8环境。
```
Comment (ZHUI) [inline llm/gpt-3/README.md]:
```suggestion
2. `use_flash_attention` 需要在A100机器开启，否则loss可能不正常（很快变成0.00x,非常小不正常）。建议使用cuda11.8环境。
```

File: llm/gpt-3/README.md [hunk 8]
```diff
@@ -28,15 +30,20 @@ mv gpt_en_dataset_300m_ids.npy ./data
...
 ```shell
+task_name="gpt3_hybid"
 export PYTHONPATH="../../PaddleNLP/"
 export FLAGS_cudnn_deterministic=True
 log_dir="log"
 rm -rf $log_dir
 
 python -u  -m paddle.distributed.launch \
-    --gpus "0" \
+    --gpus "6,7" \
```
Comment (ZHUI) [inline llm/gpt-3/README.md]:
```suggestion
    --gpus "0" \
```

File: llm/gpt-3/modeling.py [hunk 9]
```diff
@@ -96,109 +96,72 @@ class MultiHeadAttention(nn.Layer):
...
+        self.enable_recompute = False
+
+        config.use_flash_attention = config.use_flash_attention if flash_attention else None
+
+        self.head_dim = config.hidden_size // config.num_attention_heads
+        assert self.head_dim * config.num_attention_heads == config.hidden_size, "hidden_size must be divisible by num_attention_heads"
+
+        if config.tensor_parallel_degree > 1:
+            assert config.num_attention_heads % config.tensor_parallel_degree == 0
+            config.num_attention_heads = config.num_attention_heads // config.tensor_parallel_degree
```
Comment (ZHUI) [inline llm/gpt-3/modeling.py]:
```suggestion
            self.num_attention_heads = config.num_attention_heads // config.tensor_parallel_degree
```
修改了原始变量的地方，建议重新赋值一遍。不要直接修改config

File: llm/gpt-3/modeling.py [hunk 10]
```diff
@@ -270,10 +233,10 @@ def gen_cache(self, key, value=None, type=Cache):
             return self.StaticCache(k, v)
         elif value is None:  # incremental_state
             k = layers.fill_constant_batch_size_like(
-                input=key, shape=[-1, self.num_heads, 0, self.head_dim], dtype=key.dtype, value=0
+                input=key, shape=[-1, self.config.num_attention_heads, 0, self.head_dim], dtype=key.dtype, value=0
```
Comment (ZHUI) [inline llm/gpt-3/modeling.py]:
```suggestion
                input=key, shape=[-1, self.num_attention_heads, 0, self.head_dim], dtype=key.dtype, value=0
```

File: llm/gpt-3/modeling.py [hunk 11]
```diff
@@ -270,10 +233,10 @@ def gen_cache(self, key, value=None, type=Cache):
...
             )
             v = layers.fill_constant_batch_size_like(
-                input=key, shape=[-1, self.num_heads, 0, self.head_dim], dtype=key.dtype, value=0
+                input=key, shape=[-1, self.config.num_attention_heads, 0, self.head_dim], dtype=key.dtype, value=0
```
Comment (ZHUI) [inline llm/gpt-3/modeling.py]:
```suggestion
                input=key, shape=[-1, self.num_attention_heads, 0, self.head_dim], dtype=key.dtype, value=0
```

File: llm/gpt-3/modeling.py [hunk 12]
```diff
@@ -96,109 +96,72 @@ class MultiHeadAttention(nn.Layer):
...
-            if self.fuse_attention_qkv:
-                assert self.kdim == embed_dim, "embed_dim should be equal to kdim"
-                assert self.vdim == embed_dim, "embed_dim should be equal to vidm"
 
+        self.config = config
+
+        # Recompute defaults to False and is controlled by Trainer
+        self.enable_recompute = False
+
+        config.use_flash_attention = config.use_flash_attention if flash_attention else None
```
Comment (ZHUI) [inline llm/gpt-3/modeling.py]:
```suggestion
        self.use_flash_attention = config.use_flash_attention if flash_attention else None
```

File: llm/gpt-3/modeling.py [hunk 13]
```diff
@@ -311,20 +274,20 @@ def core_attn(self, q, k, v, attn_mask=None):
...
+                weights = F.dropout(weights, self.config.hidden_dropout_prob, training=self.training, mode="upscale_in_train")
 
         out = paddle.matmul(weights, v)
 
         # combine heads
         out = tensor.transpose(out, perm=[0, 2, 1, 3])
         out = tensor.reshape(x=out, shape=[0, 0, -1])
 
-        return (out, weights) if self.need_weights else out
+        return (out, weights) if self.config.need_weights else out
```
Comment (ZHUI) [inline llm/gpt-3/modeling.py]: https://github.com/huggingface/transformers/blob/main/src/transformers/models/gpt2/modeling_gpt2.py#L299C9-L299C26

参考此处，`self.config.need_weights` 换成，forward 函数中的 output_attentions 参数。

**Review by ZHUI — APPROVED** [review]
LGTM

**Pull request merged**